### DISCUSSION TITLE
Suggested change in Template

### DISCUSSION THREAD

**Discussion by drjdpowell** [NEVSTOP-LAB/.github#20] [Ideas]
The first impression a new User gets of your template is the large number of wires coming out of the "Parse State Queue++".  

![image](https://github.com/user-attachments/assets/c5eefb3c-bd43-48e5-908f-fd755ae1926d)

It gives the impression that this framework is going to be hard to understand.  It looks complex.  Here is my suggestion of moving things around a bit:

![image](https://github.com/user-attachments/assets/daea8a3c-3016-4ad2-b653-cb25b088ad94)

I've 

- separated the various Call/Response stuff from the original JKI-SM stuff
- made it obvious what the "Last State Response" is 
- tries to use different names to distinguish between my response to someone else versus someone else's response to me (which I found very confusing at first)

**Comment by drjdpowell** [comment]
Oh, I also
- Made Init State a constant, as one wouldn't normally change that
- stopped passing the "My Response Args" into the case structure.   This make it clear that we are sending them off in our response.  I'm not sure why one would want to keep a copy of them, but the user can easily do that if they like.

**Comment by nevstop** [comment]
Great suggestion! It will be included in the coming release.

**Comment by nevstop** [comment]
> Oh, I also
> 
> * Made Init State a constant, as one wouldn't normally change that
> * stopped passing the "My Response Args" into the case structure.   This make it clear that we are sending them off in our response.  I'm not sure why one would want to keep a copy of them, but the user can easily do that if they like.

Here is why I passed the "My Response Args" in to case structure. If caller send sync message of "Macro: DoSomething" to callee, which is consist of several sub-states. This wire is needed to pass the actual return value from the middle state in "Action： Step3".

``` c
\\ sub-states of "Macro: DoSomething"
Action: Step1
Action: Step2 \\ return value
Action: Step3
```

Caller will get the return value as response for "Macro: DoSomething" after all sub-states are executed by callee.

But anyway, I agree with you that the users could easily do it by themselves. That's a corner case. I accept the change. It helps the template looks better.  :)

**Comment by drjdpowell** [comment]
Idea: have "Parse State Queue" use a feedback node to save the last non-empty "My Response Args".  Then Step 3 won't overwrite the return value provided by Step 2.

**Comment by nevstop** [comment]
> Idea: have "Parse State Queue" use a feedback node to save the last non-empty "My Response Args". Then Step 3 won't overwrite the return value provided by Step 2.

I think it would be more appropriate to leave this decision to the users. If we do it within the framework, it may limit users' ability to modify the behavior as needed. Additionally, it could overwrite some expected empty responses.

**Comment by nevstop** [comment]
@drjdpowell It Looks better now. This could be a picture for a spot the difference game.  :)
![image](https://github.com/user-attachments/assets/3d26c55b-b9c5-47c6-b3ad-f8e3a6549bdf)
